src/main.rs

```diff
@@ -133,14 +176,78 @@ fn test(matches: clap::ArgMatches) {
         }
     };
 
-    let runner_start = Runner::<Mdl, TensorAnalysis, ()>::default().with_expr(&start);
-    runner_start
-        .egraph
-        .dot()
-        .to_svg("target/start.svg")
-        .unwrap();
-    let time_start = get_full_graph_runtime(&runner_start);
-    println!("Start graph runtime: {}", time_start);
+    // Get multi-pattern rules. learned_rules are the learned rules from TASO,
+    // pre_defined_multi are the hand-specified rules from TASO
+    let multi_patterns = if let Some(rule_file) = matches.value_of("multi_rules") {
+        let learned_rules =
+            read_to_string(rule_file).expect("Something went wrong reading the rule file");
+        let pre_defined_multi = PRE_DEFINED_MULTI.iter().map(|&x| x);
+        let multi_rules: Vec<&str> = learned_rules.split("\n").chain(pre_defined_multi).collect();
+        MultiPatterns::with_rules(multi_rules)
+    } else {
+        let multi_rules: Vec<&str> = PRE_DEFINED_MULTI.iter().map(|&x| x).collect();
+        MultiPatterns::with_rules(multi_rules)
+    };
+
+    // Run saturation
+    let n_sec = matches
+        .value_of("n_sec")
+        .map_or(10, |s| s.parse::<u64>().unwrap());
+    let time_limit_sec = Duration::new(n_sec, 0);
+    let iter_limit = matches
+        .value_of("n_iter")
+        .map_or(1, |s| s.parse::<usize>().unwrap());
+
+    let runner = if use_multi {
+        // This hook function (which applies the multi-pattern rules) will be called at the
+        // beginning of each iteration in equality saturation
+        Runner::<Mdl, TensorAnalysis, ()>::default()
+            .with_node_limit(100000)
```

Comment:
Does this limit number of nodes in e-graph?

Reply:
Yes, right now I am not relying on this limit, so was just setting to a large number